assets/src/hooks/countdown_timer.ts

```diff
@@ -1,68 +1,107 @@
 export const CountdownTimer = {
   mounted() {
-    const { timerId, submitButtonId, timeOutInMins, startTimeInMs, effectiveTimeInMs, autoSubmit } =
-      this.el.dataset;
-    const parsedTimeOutInMins = parseInt(timeOutInMins, 10);
+    const {
+      timerId,
+      submitButtonId,
+      timeOutInMins,
+      startTimeInMs,
+      effectiveTimeInMs,
+      gracePeriodInMins,
+      autoSubmit,
+    } = this.el.dataset;
+
+    // Parse the dataset values
+    const parsedTimeOutInMs = parseInt(timeOutInMins, 10) * 60 * 1000;
     const parsedStartTimeInMs = parseInt(startTimeInMs, 10);
     const parsedEffectiveTimeInMs = parseInt(effectiveTimeInMs, 10);
+    const parsedGracePeriodInMs = parseInt(gracePeriodInMins, 10) * 60 * 1000;
     const parsedAutoSubmit = autoSubmit === 'true';
 
-    countdownTimer(
-      timerId,
-      submitButtonId,
-      parsedTimeOutInMins,
-      parsedStartTimeInMs,
-      parsedEffectiveTimeInMs,
-      parsedAutoSubmit,
-    );
-    window.addEventListener('beforeunload', this.handleBeforeUnload.bind(this, submitButtonId));
+    this.isPageHidden = false;
+    this.submitButtonId = submitButtonId;
+
+    const currentTimeInMs = Date.now();
+
+    // Check if the current time is greater than or equal to the end time plus the grace period
+    if (currentTimeInMs >= parsedStartTimeInMs + parsedEffectiveTimeInMs + parsedGracePeriodInMs) {
+      this.submitForm();
+    } else {
+      // Start the countdown timer if it hasn't expired yet
+      countdownTimer(
+        timerId,
+        submitButtonId,
+        parsedTimeOutInMs,
+        parsedStartTimeInMs,
+        parsedEffectiveTimeInMs,
+        parsedGracePeriodInMs,
+        parsedAutoSubmit,
+      );
+    }
+
+    // Add event listeners to handle visibility changes and page unload
+    document.addEventListener('visibilitychange', this.handleVisibilityChange.bind(this));
+    window.addEventListener('beforeunload', this.handleBeforeUnload.bind(this));
   },
   destroyed() {
+    // Remove event listeners when the component is destroyed
+    document.removeEventListener('visibilitychange', this.handleVisibilityChange.bind(this));
     window.removeEventListener('beforeunload', this.handleBeforeUnload.bind(this));
   },
-  handleBeforeUnload(submitButtonId: string) {
-    const submitButton = document.getElementById(submitButtonId);
-    submitButton ? submitButton.click() : console.error('Submit button not found');
+  handleVisibilityChange() {
+    this.isPageHidden = document.hidden;
+  },
+  handleBeforeUnload() {
+    if (this.isPageHidden) {
```

Comment:
This logic seems to indicate that this will only be auto submitted when the page is closed and the page was hidden, is that the correct behavior?

I would have expected this to instead check if time is expired and submit if so, regardless of page visibility state.

Or even better - do nothing when page visibility or unload happens and let the backend handle the auto submit if a student leaves. There is already logic here to handle when the page is loaded and past due to auto submit, so that should handle the case of reloading the page past the due date

Reply:
Yes, there already is backend code that will auto submit an assessment after the effective deadline (with about 1 minute of added slack).  

The code in the frontend should ONLY auto submit when the effective deadline is hit and the page is still visible.

The changes that need to be made here are to remove all of the event handlers for `beforeUnload` and `handleVisibilityChange` and then ONLY auto submit from the front end when the effective deadline is reached.  
Also remove the code on initialization which checks the effective deadline and auto submits.

Reply:
Ok, I got it. I will remove all references to `beforeUnload` and `handleVisibilityChange ` (in the `mounted` should only call the function with its corresponding parameters). Now I wonder if it makes sense the logic I built into the `countdownTimer `function to take into account not only the `effective time`, but also the `grace period` time.